### PR TITLE
feat(pageserver): support dry-run for gc-compaction, add statistics

### PR DESCRIPTION
**Pull request by skyzh**
## Problem

Add dry-run mode that does not produce any image layer + delta layer. I will use this code to do some experiments and see how much space we can reclaim for tenants on staging. Part of https://github.com/neondatabase/neon/issues/8002

## Summary of changes

* Add dry-run mode that runs the full compaction process without updating the layer map. (We never call finish on the writers and the files will be removed before exiting the function).
* Add compaction statistics and print them at the end of compaction.

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: pageserver/src/tenant.rs [hunk 1]
```diff
@@ -7898,11 +7910,34 @@ mod tests {
...
         let cancel = CancellationToken::new();
-        tline.compact_with_gc(&cancel, &ctx).await.unwrap();
+
+        tline
+            .compact_with_gc(
+                &cancel,
+                {
+                    let mut flags = EnumSet::new();
+                    flags.insert(CompactFlags::DryRun);
+                    flags
```
Comment (arpad-m) [inline pageserver/src/tenant.rs:7920]: I think it's better to have this declared outside like:

```Rust
let cancel = ...
let flags = ...
flags.insert(...);
timeline.compact_with_gc(...)
```

File: pageserver/src/tenant/timeline/compaction.rs [hunk 2]
```diff
@@ -82,31 +86,118 @@ impl KeyHistoryRetention {
...
+            stat.produce_key(&val);
             delta_writer.push((key, lsn, val));
         }
         Ok(())
     }
 }
 
+#[derive(Debug, Serialize, Default)]
+pub struct CompactionStatistics {
+    num_delta_layer_visited: usize,
```
Comment (arpad-m) [inline pageserver/src/tenant/timeline/compaction.rs:120]: Maybe a `NumSize{ num: u64, size: u64, }` type would be better? Then we'd have half of the number of variables in this struct.

Also, in theory we can process more data than fits in memory so u64 is better than usize (we dont have 32 bit builds but still).

**Review by arpad-m — APPROVED** [review]
I'm not 100% sure an EnumSet is the best way forward, but I guess adding a trait like with tiered compaction is too complicated. It's fine I guess.

File: pageserver/src/tenant/timeline/compaction.rs (hunk 2 above)
Comment (skyzh) [reply pageserver/src/tenant/timeline/compaction.rs]: resolved in https://github.com/neondatabase/neon/pull/8557/commits/dffe43384bd21dd986c28cafcd8faba243b8179a

File: pageserver/src/tenant.rs (hunk 1 above)
Comment (skyzh) [reply pageserver/src/tenant.rs]: resolved in https://github.com/neondatabase/neon/pull/8557/commits/f535877ab40fe5ceb1ca856cd86a85acb5901dbb

**Review by arpad-m — APPROVED** [review]

**Pull request merged**